### PR TITLE
DISCUSSION-23: Allow to chose the storage location by "application"

### PR DESCRIPTION
**Pull request by surli**
## JIRA
https://jira.xwiki.org/browse/DISCUSSION-23

## Description
This commit starts a big refactoring of Discussion to integrate the
concept of applicationHint: the idea is to be able to manipulate
discussion with a dedicated applicationHint so that dedicated
applications can be built on top of it. Right now this hint is only used
for choosing the location where to store the discussions, but we could
imagine other usecases later with this hint.

### CODE REVIEW

File: discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionContextService.java [hunk 1]
```diff
@@ -55,12 +58,11 @@
      *
      * @param name the name
      * @param description the description
-     * @param referenceType the reference type
      * @param entityReference the entity reference
      * @return the found or created discussion context
      */
-    Optional<DiscussionContext> getOrCreate(String name, String description, String referenceType,
-        String entityReference);
+    Optional<DiscussionContext> getOrCreate(String applicationHint, String name, String description,
```
Comment (manuelleduc) [inline discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionContextService.java:66]: Missing javadoc for applicationHint

File: discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.contrib.discussions;
+
+import org.xwiki.contrib.discussions.domain.references.AbstractDiscussionReference;
+
+public interface DiscussionReferencesResolver
```
Comment (manuelleduc) [inline discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java:34]: Missing javadoc

File: discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+@Unstable
+@Role
+public interface DiscussionReferencesResolver
+{
+    /**
+     * Perform the resolution of a given serialized reference to the given {@link AbstractDiscussionReference} concrete
+     * type. Note that the serialization might not contain any information about the actual type, so it might not be
+     * possible to check that the type is correct for this reference.
+     *
+     * @param serializedReference a serialization of an {@link AbstractDiscussionReference}.
```
Comment (manuelleduc) [inline discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java]: Shouldn't finish with a dot

File: discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesSerializer.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+@Role
+public interface DiscussionReferencesSerializer
+{
+    /**
+     * Serialize the given reference to store it as a string.
+     *
+     * @param discussionReference the reference to serialize.
+     * @return a serialization of the reference.
+     */
+    String serialize(AbstractDiscussionReference discussionReference);
```
Comment (manuelleduc) [inline discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesSerializer.java:42]: Why not one method per kind of reference? Unless we plan to make it open to others to provide their own reference kinds?

File: discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+     * @param type the concrete type to obtain
+     * @param <T> the actual given concrete type
+     * @return an instance of the given type with the information of the reference.
+     */
+    <T extends AbstractDiscussionReference> T resolve(String serializedReference, Class<T> type);
```
Comment (manuelleduc) [inline discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java:46]: Might be interesting to add resolveXXX method, one for each kind of serialized reference kind. Could be useful when we already know statically the kind of the reference to resolve.

File: discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionsActorServiceResolver.java [hunk 6]
```diff
@@ -40,5 +40,5 @@
      * @param type the type
      * @return the actor service, {@link Optional#empty()} if the resolution failed
      */
-    Optional<DiscussionsActorService> get(String type);
+    DiscussionsActorService get(String type);
```
Comment (manuelleduc) [inline discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionsActorServiceResolver.java:42]: (small detail) Why is the Optional removed? I agree the default implementation does not return the empty value, but could be the case for another implementation.

File: discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java (hunk 5 above)
Comment (manuelleduc) [inline discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java:46]: Why not Optional<T> as the resolution could fail?

File: discussions-default/src/main/java/org/xwiki/contrib/discussions/internal/DefaultDiscussionReferencesSerializer.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+@Component
+@Singleton
+public class DefaultDiscussionReferencesSerializer implements DiscussionReferencesSerializer
+{
+    @Override
+    public String serialize(AbstractDiscussionReference discussionReference)
+    {
+        String result = discussionReference.getReference();
+        if (!StringUtils.isEmpty(discussionReference.getApplicationHint())) {
+            result = String.format("%s;applicationHint=%s", result, discussionReference.getApplicationHint());
```
Comment (manuelleduc) [inline discussions-default/src/main/java/org/xwiki/contrib/discussions/internal/DefaultDiscussionReferencesSerializer.java:48]: We need to discuss what's the best serialization format.

File: discussions-default/src/main/java/org/xwiki/contrib/discussions/internal/DefaultDiscussionsRightService.java [hunk 8]
```diff
@@ -84,6 +84,7 @@ public void initialize() throws InitializationException
         }
     }
 
+
```
Comment (manuelleduc) [inline discussions-default/src/main/java/org/xwiki/contrib/discussions/internal/DefaultDiscussionsRightService.java]: useless change ;)

File: discussions-rest/discussions-rest-default/src/checkstyle/checkstyle-suppressions.xml [hunk 9]
```diff
@@ -26,4 +26,5 @@
 
 <suppressions>
   <suppress checks="ParameterNumber" files="DefaultDiscussionREST.java"/>
+  <suppress checks="ClassFanOutComplexity" files="DefaultDiscussionREST.java"/>
```
Comment (manuelleduc) [inline discussions-rest/discussions-rest-default/src/checkstyle/checkstyle-suppressions.xml:29]: We'll need to work on that on a second time :)

**Review by manuelleduc — COMMENTED** [review]
Partial review, mainly notes for myself.

File: discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesSerializer.java (hunk 4 above)
Comment (surli) [reply discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesSerializer.java:42]: So for the serialization it's actually not needed, you can serialize all the `AbstractDiscussionReference` the same way, so I don't see the point of adding complexity to chose the right method for serializing.

File: discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java (hunk 5 above)
Comment (surli) [reply discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java:46]: Personally I think it's enough to have a single resolve method taking the type.

File: discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionsActorServiceResolver.java (hunk 6 above)
Comment (surli) [reply discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionsActorServiceResolver.java:42]: So I improved a bit the javadoc there to specify that the contract of this resolver should always fallback on a default service if a specific implementation for the type does not exist. So it should never be the case to return null on another implementation.

File: discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java (hunk 5 above)
Comment (surli) [reply discussions-api/src/main/java/org/xwiki/contrib/discussions/DiscussionReferencesResolver.java:46]: The resolution could only fail if someone provides a new kind of AbstractDiscussionReference and don't provide an appropriate constructor. So in practice it should never happen, and it wouldn't be the responsibility of the API consumer to be aware of that. For the consumer of the API the reference should always be resolved.

File: discussions-default/src/main/java/org/xwiki/contrib/discussions/internal/DefaultDiscussionReferencesSerializer.java (hunk 7 above)
Comment (surli) [reply discussions-default/src/main/java/org/xwiki/contrib/discussions/internal/DefaultDiscussionReferencesSerializer.java:48]: So as detailed in the javadoc, the format is based on XWiki Page reference format which accepts parameters key/value, see: https://extensions.xwiki.org/xwiki/bin/view/Extension/Model%20Module#HPageReferences

**Pull request merged**